### PR TITLE
Chain builtin auth fixed

### PR DESCRIPTION
**Pull request by frafra**
Updated/fixed version of #4248 and #4575, rebased onto master.

### Features:

- [X] includes tests covering changes
- [ ] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes API changes
- [ ] includes bugfix for possible backport

### CODE REVIEW

File: ckan/authz.py [hunk 1]
```diff
@@ -102,13 +102,21 @@ def _build(self):
...
-                fetched_auth_functions[name] = functools.partial(func, prev_func)
+                if name in fetched_auth_functions:
+                    prev_func = fetched_auth_functions[name]
+                else:
+                    # fallback to chaining off the builtin auth function
+                    prev_func = self._functions[name]
+                fetched_auth_functions[name] =\
+                    functools.partial(func, prev_func)
+                new_name = func.__name__ or prev_func.__name__
+                fetched_auth_functions[name].__name__ = new_name
```
[the remark this answers is not in the document]
Comment (frafra) [reply ckan/authz.py]: Another option would be to remove/change `auth_function.__name__` in `ckan/authz.py`.

[the remark this answers is not in the document]
Comment (wardi) [reply ckan/authz.py]: right, can't we change it to `action` in the error message?

[the remark this answers is not in the document]
Comment (frafra) [reply ckan/authz.py]: Sure we can :-)
"Action action requires an authenticated user" seems a bit odd to me, so I set it to "(unnamed)", but let me know if there is something more appropriate that can be used.

[the remark this answers is not in the document]
Comment (wardi) [reply ckan/authz.py]: I mean there's a variable called `action` that should have the action name in it

[the remark this answers is not in the document]
Comment (frafra) [reply ckan/authz.py]: Uh, right.

**Comment by wardi** [comment]
just the minor style fix requested from Adrià https://github.com/ckan/ckan/pull/4248#discussion_r190902584 and a small test and this is good to go

**Comment by frafra** [comment]
A test seems the only thing that is missing now.

**Comment by kmbn** [comment]
@frafra Just to clarify the workflow—you're still working on the test that will complete your PR, correct?

**Comment by frafra** [comment]
@kmbn I am writing the test right now.

**Comment by frafra** [comment]
The simple patch I wrote fails for current master, while it is OK on this branch containing the fix.

**Comment by frafra** [comment]
@wardi it seems fine to me; is there anything else to do before merging it?

**Pull request merged**